Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -840,23 +844,33 @@ extension TaskHandler: ChannelDuplexHandler {
             self.writeBody(request: request, context: context)
         }.flatMap {
             context.eventLoop.assertInEventLoop()
-            if case .endOrError = self.state {
+
+            switch self.state {
+            case .idle:
+                // since this code path is called from `write` and write sets state to sendingBody
+                preconditionFailure("should not happen")
+            case .sendingBodyWaitingResponseHead:
+                self.state = .bodySentWaitingResponseHead
+            case .sendingBodyResponseHeadReceived:
+                self.state = .bodySentResponseHeadReceived
+            case .bodySentWaitingResponseHead, .bodySentResponseHeadReceived:
+                preconditionFailure("should not happen")
+            case .endOrError, .redirected:
```

Comment:
mind adding a comment why that's okay in `.endOrError`, or `.redirected`? I don't quite get how we can be in `.end` _before_ we get the end of the body.

Reply:
(similar for `.redirected`)

Reply:
rewrote, added a test